File: app/src/main/java/com/insomniacgks/newmoviesandshows/data/GetReviews.java
package com.insomniacgks.newmoviesandshows.data;

import android.content.Context;
import android.os.AsyncTask;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatDelegate;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.insomniacgks.newmoviesandshows.R;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.ArrayList;

public class GetReviews extends AsyncTask<String, Void, ArrayList<String[]>> {
    private Context context;
    private int id;
    private RecyclerView recyclerView;
    private RelativeLayout rl;
    private String type;

    public GetReviews(Context context, int id, RecyclerView recyclerView, RelativeLayout rl, String type) {
        this.context = context;
        this.id = id;
        this.recyclerView = recyclerView;
        this.rl = rl;
        this.type = type;
    }

    protected ArrayList<String[]> doInBackground(String... strings) {
        ArrayList<String[]> reviews = null;
        try {
            HttpURLConnection client = (HttpURLConnection) new URL("http://api.themoviedb.org/3/" + type + "/" + id + "/reviews?api_key=" + Constants.API_KEY + "&language=en-US").openConnection();
            client.setRequestMethod("GET");
            client.connect();
            BufferedReader reader = new BufferedReader(new InputStreamReader(client.getInputStream()));
            StringBuilder builder = new StringBuilder();
            while (true) {
                String line = reader.readLine();
                if (line == null) {
                    break;
                }
                builder.append(line);
            }
            reviews = new ArrayList<>();
            JSONArray resultArray = new JSONObject(builder.toString()).getJSONArray("results");
            int i = 0;
            while (i < resultArray.length()) {
                JSONObject object = resultArray.getJSONObject(i);
                reviews.add(new String[]{object.getString("author"), object.getString("content")});
                i++;
            }
        } catch (Exception ignored) {
        }
        return reviews;
    }

    protected void onPostExecute(ArrayList<String[]> strings) {
        super.onPostExecute(strings);
        if (strings != null && strings.size() != 0) {
            RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(this.context);
            ReviewsRecyclerViewAdapter myAdapter = new ReviewsRecyclerViewAdapter(this.context, strings);
            recyclerView.setLayoutManager(layoutManager);
            recyclerView.setAdapter(myAdapter);
        } else {
            rl.setVisibility(View.GONE);
        }
    }


    class ReviewsRecyclerViewAdapter extends RecyclerView.Adapter<ReviewsRecyclerViewAdapter.ViewHolder> {
        private ArrayList<String[]> reviewss;
        private Context context;

        ReviewsRecyclerViewAdapter(Context context, ArrayList<String[]> Reviewss) {
            this.context = context;
            this.reviewss = Reviewss;
        }


        @NonNull
        @Override
        public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            AppCompatDelegate.setCompatVectorFromResourcesEnabled(true);
            return new ViewHolder(LayoutInflater.
                    from(parent.getContext()).inflate(R.layout.reviews_card_layout, parent, false));
        }

        @Override
        public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
            holder.author.setText(String.format("Review By %s:", this.reviewss.get(position)[0]));
            holder.content.setText(this.reviewss.get(position)[1]);
        }

        @Override
        public int getItemCount() {
            return reviewss.size();
        }

        class ViewHolder extends android.support.v7.widget.RecyclerView.ViewHolder {

            TextView content, author;

            ViewHolder(View itemView) {
                super(itemView);
                this.author = itemView.findViewById(R.id.review_author);
                this.content = itemView.findViewById(R.id.review_content);
            }

        }
    }
}

File: app/src/main/java/com/insomniacgks/newmoviesandshows/data/GetReviews.kt
package com.insomniacgks.newmoviesandshows.data

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Color
import android.os.AsyncTask
import android.preference.PreferenceManager
import android.support.v4.content.ContextCompat
import android.support.v7.app.AppCompatDelegate
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.RelativeLayout
import android.widget.TextView
import com.insomniacgks.newmoviesandshows.R
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL
import java.util.*

class GetReviews(@field:SuppressLint("StaticFieldLeak") private val context: Context, private val id: Int, @field:SuppressLint("StaticFieldLeak") private val recyclerView: RecyclerView, @field:SuppressLint("StaticFieldLeak") private val rl: RelativeLayout, private val type: String) : AsyncTask<String, Void, ArrayList<Array<String>>>() {

    override fun doInBackground(vararg strings: String): ArrayList<Array<String>>? {
        var reviews: ArrayList<Array<String>>? = null
        try {
            val client = URL("http://api.themoviedb.org/3/" + type + "/" + id + "/reviews?api_key=" + Constants.API_KEY + "&language=en-US").openConnection() as HttpURLConnection
            client.requestMethod = "GET"
            client.connect()
            val reader = BufferedReader(InputStreamReader(client.inputStream))
            val builder = StringBuilder()
            while (true) {
                val line = reader.readLine() ?: break
                builder.append(line)
            }
            reviews = ArrayList()
            val resultArray = JSONObject(builder.toString()).getJSONArray("results")
            var i = 0
            while (i < resultArray.length()) {
                val `object` = resultArray.getJSONObject(i)
                reviews.add(arrayOf(`object`.getString("author"), `object`.getString("content")))
                i++
            }
        } catch (ignored: Exception) {
        }

        return reviews
    }

    override fun onPostExecute(strings: ArrayList<Array<String>>?) {
        super.onPostExecute(strings)
        if (strings != null && strings.size != 0) {
            val layoutManager = LinearLayoutManager(this.context)
            val myAdapter = ReviewsRecyclerViewAdapter(this.context, strings)
            recyclerView.layoutManager = layoutManager
            recyclerView.adapter = myAdapter
        } else {
            rl.visibility = View.GONE
        }
    }


    internal inner class ReviewsRecyclerViewAdapter(private val context: Context, private val reviewss: ArrayList<Array<String>>) : RecyclerView.Adapter<ReviewsRecyclerViewAdapter.ViewHolder>() {
        private val isDarkModeOn: Boolean = PreferenceManager.getDefaultSharedPreferences(context).getBoolean("dark_mode", false)

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
            AppCompatDelegate.setCompatVectorFromResourcesEnabled(true)
            return ViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.reviews_card_layout, parent, false))
        }

        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            if (isDarkModeOn) {
                holder.author.setTextColor(Color.WHITE)
                holder.content.setTextColor(Color.WHITE)
                holder.LayoutBack.setBackgroundColor(ContextCompat.getColor(this.context, R.color.black_theme_color))
            }
            holder.author.text = String.format("Review By %s:", this.reviewss[position][0])
            holder.content.text = this.reviewss[position][1]
        }

        override fun getItemCount(): Int {
            return reviewss.size
        }

        internal inner class ViewHolder(itemView: View) : android.support.v7.widget.RecyclerView.ViewHolder(itemView) {

            var content: TextView
            var author: TextView
            var LayoutBack: LinearLayout

            init {
                this.LayoutBack = itemView.findViewById(R.id.layout_back)
                this.author = itemView.findViewById(R.id.review_author)
                this.content = itemView.findViewById(R.id.review_content)
            }

        }
    }
}